design/ceph/object/ceph-cosi-driver.md

```diff
@@ -0,0 +1,87 @@
+
+### Pre-requisites
+
+- COSI CRDs should be installed in the cluster via following command
+
+```bash
+kubectl apply -k github.com/kubernetes-sigs/container-object-storage-interface-api
+```
+
+- COSI controller should be deployed in the cluster via following command
+
+```bash
+kubectl apply -k github.com/kubernetes-sigs/container-object-storage-interface-controller
+```
+
+- Create Service Account and RBAC for COSI driver
+
+```bash
+kubectl create -f https://github.com/ceph/ceph-cosi/blob/master/resources/sa.yaml
+kubectl create -f https://github.com/ceph/ceph-cosi/blob/master/resources/rbac.yaml
+```
+
+### Changes in Operator.yaml
+
+Below option validates whether the COSI controller is deployed in the cluster. If it is not deployed, it will log an warning message and when user tries to create object store woth cosi, it will reconcile until the controller is deployed.
+
+```
+  cosiController:
+    enabled: true
+```
+
+### Changes in ObjectStore.yaml
+
+Following option will bring up the Ceph COSI driver for the RGW server. It will also create `CephObjectStoreUser`  objectstorage-provisioner which provides credentials for Ceph COSI driver.
+```
+  cephCosiDrvierSpec:
+    enabled: true
+```
+
+### How Rook can improve COSI driver reliability
+
+Rook can ensure that resources need for Ceph COSI driver such as Ceph object store is deployed and running in the cluster before creating requests for Bucket and its Access. Rook can also ensure that the COSI driver is deployed with the correct RBAC permissions, with the correct version of COSI controller. Rook should also prevent deletion of the ceph object Store if there are any buckets or bucket access in the cluster which already present with current codebase. Rook also need to provide secret containing the credentials for the ceph object store to the COSI driver. The secret need to updated if the ceph object store endpoint changes or credentials are changed. Rook need to bring up multiple drivers for multiple ceph object stores in the cluster. Rook also need to support update of COSI-Controller and COSI-Driver images independently
+
+### Coexistence of COSI and libbucket provisioner
+
+Currently the ceph object store provisioned via Object Bucket Claim (OBC). They both can co exist and can even use same backend bucket from ceph storage. No deployment/configuration changes are required to support both. The libbucket probvisioner is deprecated and eventually will be replaced by COSI when it becomes more and more stable. The CRDs used by both are different hence there is no conflicts between them.
+
+### Ceph COSI Driver Requirements
+
+- Ceph Object Store should be deployed and running in the cluster
+- The credentials/endpoint for the ceph object store should be available by creating ceph object store user with proper permissions
+- The COSI controller should be deployed in the cluster
+- Rook can able to manage multiple ceph cosi drivers
```

Comment:
I'm still not clear... Why do we need multiple COSI drivers? Or can we just support one COSI driver, just like we do for CSI?

Reply:
Yes thats based on last discussion. Currently, Ceph COSI driver only supports one endpoint, We can have one single driver for every ceph object stores in Rook. Since ceph cosi driver is light weight we thought of having multiple drivers than a single one. But we can change what ever makes good scalable design

Reply:
Could you add a section that lists which resources/components/pods will be created by Rook (or not included with Rook)? It will help understand the single COSI controller, the object-store-specific COSI driver, etc.

Reply:
The latest discussion was that there should be one Ceph COSI driver per k8s cluster. Let's adjust this design to reflect that.